atef/bin/check.py

```diff
@@ -256,18 +264,55 @@ def log_results_rich(
     console.print(root)
 
 
-def check_and_log(
+async def check_and_log(
     config: AnyConfiguration,
     console: rich.console.Console,
     verbose: int = 0,
     client: Optional[happi.Client] = None,
     name_filter: Optional[Sequence[str]] = None,
+    parallel: bool = True,
+    cache: Optional[DataCache] = None,
 ):
-    """Check a configuration and log the results."""
+    """
+    Check a configuration and log the results.
+
+    Parameters
+    ----------
+    config : AnyConfiguration
+        The configuration to check.
+    console : rich.console.Console
+        The rich console to write output to.
+    verbose : int, optional
+        The verbosity level for the output.
+    client : happi.Client, optional
+        The happi client, if available.
+    name_filter : Sequence[str], optional
+        A filter for names.
+    parallel : bool, optional
+        Pre-fill cache in parallel when possible.
+    cache : DataCache
+        The data cache instance.
+    """
     items = []
     name_filter = list(name_filter or [])
     severities = []
-    for prepared in PreparedComparison.from_config(config, client=client):
+
+    if cache is None:
+        cache = DataCache()
+
+    all_prepared = list(
+        PreparedComparison.from_config(config, client=client, cache=cache)
+    )
+
+    cache_fill_tasks = []
```

Comment:
It's only here because I read the docs for once:
https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task
> Important
> 
> Save a reference to the result of this function, to avoid a task disappearing mid execution. The event loop only keeps weak references to tasks. A task that isn’t referenced elsewhere may get garbage-collected at any time, even before it’s done. For reliable “fire-and-forget” background tasks, gather them in a collection:

Reply:
I thought it might be something like this... I'm surprised that this is enough, but the interaction between scoping/gc isn't always intuitive to me.

Reply:
GC definitely can be confusing.

In this specific instance, the list holds a hard reference and persists throughout the coroutine, so it - and the tasks it holds - should only be gc'd after it returns (or raises)